### PR TITLE
Allow direct use of OPTE/XDE ports under Viona

### PR DESCRIPTION
**Pull request by FelixMcFelix**
This PR allows propolis to construct viona instances on top of `misc`-type links (i.e., OPTE pseudo-devices), in addition to the existing VNIC support.

Unfortunately, `misc` devices are not well-serviced by `dladm`. Pulling the MAC address of a misc-type device requires that we directly interface with libdllink/libdladm -- what I have here works for creating and using the network from VMs, but if there's a simpler way I'd prefer we use that.

### CODE REVIEW

File: crates/dladm/src/lib.rs [hunk 1]
```diff
@@ -115,6 +122,68 @@ impl Handle {
...
+            arg: *mut c_void,
+            macaddr: *mut sys::dladm_macaddr_attr_t,
+        ) -> sys::boolean_t {
+            let state = &mut *(arg as *mut Arg);
+            state.n_seen += 1;
+
+            if !state.written && (*macaddr).ma_addrlen == (ETHERADDRL as u32) {
+                state.mac.copy_from_slice(&(*macaddr).ma_addr[..ETHERADDRL]);
+                state.written = true;
+            }
```
[the remark this answers is not in the document]
Comment (FelixMcFelix) [reply crates/dladm/src/lib.rs:148]: Yeah, we should -- thanks!

[the remark this answers is not in the document]
Comment (FelixMcFelix) [reply crates/dladm/src/lib.rs]: Fixed in a0ca928.

File: crates/dladm/src/lib.rs [hunk 2]
```diff
@@ -115,6 +122,68 @@ impl Handle {
...
+
+        let mut state = Arg { mac, n_seen: 0, written: false };
+
+        // SAFETY: dladm_handle_t is known to be valid, and &mut reference
+        // to state is only held inside the callback.
+        Self::handle_dladm_err(unsafe {
+            sys::dladm_walk_macaddr(
+                self.inner,
+                linkid,
+                addr_of_mut!(state) as *mut c_void,
```
[the remark this answers is not in the document]
Comment (FelixMcFelix) [reply crates/dladm/src/lib.rs]: Fixed in a0ca928.

**Review by pfmooney — COMMENTED** [review]
Look good for the most part.

Since GH won't let me comment on unchanged code:
With your additions, `query_vnic()` should probably be renamed to `query_link()` or something like it.

**Comment by jclulow** [comment]
> Unfortunately, `misc` devices are not well-serviced by `dladm`. Pulling the MAC address of a misc-type device requires that we directly interface with libdllink/libdladm -- what I have here works for creating and using the network from VMs, but if there's a simpler way I'd prefer we use that.

I think we should probably just represent it as an Ethernet link?

**Review by pfmooney — APPROVED** [review]

**Comment by FelixMcFelix** [comment]
> > Unfortunately, `misc` devices are not well-serviced by `dladm`. Pulling the MAC address of a misc-type device requires that we directly interface with libdllink/libdladm -- what I have here works for creating and using the network from VMs, but if there's a simpler way I'd prefer we use that.
> 
> I think we should probably just represent it as an Ethernet link?

I've done some quick spelunking there in search of a pathway that lets us do that -- I'm assuming we get the label  according to how opte instances are registered with DLS. Today we're going via `dls_devnet_create`/`dls_devnet_destroy` from inside an ioctl at XDE. It's not clear to me yet who plumbs the 'misc' label in.

The alternatives which reference an explicit `datalink_class_t` seem to be `dls_mgmt_create`/`dls_mgmt_destroy` or `dladm_create_datalink_id`/`dladm_destroy_datalink_id`, but they seem more top down to me? I'm probably missing something here.

**Pull request merged**